### DISCUSSION TITLE
How do I efficiently find the grand MRCA in a tree?

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tskit#1490] [Q&A] [answered]
We often assume that the oldest MRCA is the root, but we can have unary nodes above the root. So what's the best way to find the grand MRCA? There are a few options: simply(keep_unary=False) then get the root, find the root and descend down until you find a non unary node, or tree.mrca(all_samples) when #1340 is addressed. I wonder what the best option is likely to be?

**Comment by jeromekelleher** [comment, marked as answer]
Unless the tree has a very large number of unary nodes above the root, then traversing down will be more efficient:
```
mrca = tree.root
while tree.num_children(mrca) == 1:
      mrca = tree.left_child(mrca)
```
(This will break in some corner cases, but should be the right option most of the time)

**Reply by jeromekelleher** [reply]
We could add a ``grand_mrca`` function to the API to do this, it's a well defined operation and could be done in C if efficiency is an issue.
